cpp/src/arrow/builder.h

```diff
@@ -682,10 +682,13 @@ class ARROW_EXPORT BinaryBuilder : public ArrayBuilder {
 
   Status Init(int64_t elements) override;
   Status Resize(int64_t capacity) override;
+  Status ReserveData(int64_t capacity);
```

Comment:
Can you add a doxygen comment, since this is a new API

I note that the `Reserve`-type methods in this header have different semantics from their STL counterparts. They are reserving _additional_ space rather than _absolute_ space (e.g. std::vector::reserve takes an absolute length as argument)

Reply:
Can we also call the argument to `ReserveData` `size` (or `elements`) instead of `capacity` (to avoid confusion about whether we are passing an incremental value vs. absolute)